src/main/java/frc/robot/subsystems/Drivetrain.java

```diff
@@ -0,0 +1,221 @@
+    private final Supplier<Double> kDRight = rightPIDNamespace.addConstantDouble("kD", 0);
+    private final Supplier<Double> waitTimeRight = rightPIDNamespace.addConstantDouble("wait time", 0);
+    private final Supplier<Double> toleranceRight = rightPIDNamespace.addConstantDouble("tolerance", 0);
+    private final PIDSettings rightPIDSettings;
+
+    private final Namespace anglePIDNamespace = namespace.addChild("angle pid");
+    private final Supplier<Double> kPAngle = anglePIDNamespace.addConstantDouble("kP", 0);
+    private final Supplier<Double> kIAngle = anglePIDNamespace.addConstantDouble("kI", 0);
+    private final Supplier<Double> kDAngle = anglePIDNamespace.addConstantDouble("kD", 0);
+    private final Supplier<Double> waitTimeAngle = anglePIDNamespace.addConstantDouble("wait time", 0);
+    private final Supplier<Double> toleranceAngle = anglePIDNamespace.addConstantDouble("tolerance", 0);
+    private final PIDSettings anglePIDSettings;
+
+    private final Namespace feedForwardNamespace = namespace.addChild("feed forward");
+    private final Supplier<Double> kS = feedForwardNamespace.addConstantDouble("kS", 0);
+    private final Supplier<Double> kV = feedForwardNamespace.addConstantDouble("kV", 0);
+    private final Supplier<Double> kA = feedForwardNamespace.addConstantDouble("kA", 0);
+    private final FeedForwardSettings feedForwardSettings;
+
+    private final Namespace trapezoidProfileNamespace = namespace.addChild("trapezoid profile settings");
+    private final Supplier<Double> maxVelocity = trapezoidProfileNamespace.addConstantDouble("max velocity", 0);
+    private final Supplier<Double> trapezoidAcceleration = trapezoidProfileNamespace.addConstantDouble
+            ("acceleration", 0);
+    private final TrapezoidProfileSettings trapezoidProfileSettings;
+
+    public static Drivetrain getInstance() {
+        if (instance == null) {
+            instance = new Drivetrain(
+                    "drivetrain",
+                    new CANSparkMax(RobotMap.CAN.DRIVETRAIN_LEFT_SPARKMAX_MASTER,
+                            CANSparkMaxLowLevel.MotorType.kBrushless),
+                    new CANSparkMax(RobotMap.CAN.DRIVETRAIN_LEFT_SPARKMAX_SLAVE,
+                            CANSparkMaxLowLevel.MotorType.kBrushless),
+                    new CANSparkMax(RobotMap.CAN.DRIVETRAIN_RIGHT_SPARKMAX_MASTER,
+                            CANSparkMaxLowLevel.MotorType.kBrushless),
+                    new CANSparkMax(RobotMap.CAN.DRIVETRAIN_RIGHT_SPARKMAX_SLAVE,
+                            CANSparkMaxLowLevel.MotorType.kBrushless));
+        }
+        return instance;
+    }
+
+    private Drivetrain(String namespaceName, CANSparkMax leftMaster, CANSparkMax leftSlave,
+                       CANSparkMax rightMaster, CANSparkMax rightSlave) {
+        super(
+                namespaceName, leftMaster, leftSlave, rightMaster, rightSlave);
+        gyro = new AHRS();
+        leftEncoder = leftMaster.getEncoder();
```

Comment:
would fit better in `getInstance()` since `leftEncoder` theoretically an indepentent component, but it would make the code uglier so do as you wish